### ISSUE TITLE
TypeScript 2.3.1 fails to compile this JS and throws TypeError

### ISSUE THREAD

**Issue by zenorbi** [Microsoft/TypeScript#15525]
**TypeScript Version:**  2.3.1 and 2.3.2

**Code**

`tsconfig.json`:
```json
{
	"compilerOptions": {
		"target": "es5",
		"allowJs": true,
		"outDir": "out",
		"rootDir": "src"
	}
}
```

`src/test.js`:
```js
someFunction(function(BaseClass) {
	class Hello extends BaseClass {
		constructor() {
			this.foo = "bar";
		}
	}
});
```

**Expected behavior:**
To compile without errors

**Actual behavior:**
throws TypeError: Cannot read property 'kind' of undefined

TypeScript 2.3.0 does not have this problem but 2.3.1 and 2.3.2 does.

**Comment by RyanCavanaugh** [comment]
```
TypeError: Cannot read property 'kind' of undefined
    at checkIndexConstraintForProperty (C:\github\TypeScript\built\local\tsc.js:43748:61)
    at C:\github\TypeScript\built\local\tsc.js:43706:21
    at Object.forEach (C:\github\TypeScript\built\local\tsc.js:1410:30)
    at checkIndexConstraints (C:\github\TypeScript\built\local\tsc.js:43704:20)
    at checkClassLikeDeclaration (C:\github\TypeScript\built\local\tsc.js:43968:17)
    at checkClassDeclaration (C:\github\TypeScript\built\local\tsc.js:43881:13)
    at checkSourceElement (C:\github\TypeScript\built\local\tsc.js:44930:28)
    at Object.forEach (C:\github\TypeScript\built\local\tsc.js:1410:30)
    at checkBlock (C:\github\TypeScript\built\local\tsc.js:42682:16)
    at checkSourceElement (C:\github\TypeScript\built\local\tsc.js:44893:28)
```

**Comment by boris-kolar** [comment]
Since this is a compiler error forcing many of us to downgrade to 2.22, can we get a fix in 2.3.3 please?

**Comment by sandersn** [comment]
The code that checks declared properties against index constraints doesn't handle javascript-style declarations (`this.foo = "bar"`) which don't have a `name` property. It needs to have a predicate for javascript declarations too.

**Comment by sandersn** [comment]
Fix is up at #15586
